File: src/test/java/jp/skypencil/javadocky/service/JavadocExtractorTest.java
package jp.skypencil.javadocky.service;

import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.File;
import java.io.IOException;
import java.nio.file.Path;
import jp.skypencil.javadocky.repository.LocalStorage;
import jp.skypencil.javadocky.repository.Storage;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;
import reactor.test.StepVerifier;

class JavadocExtractorTest {
  private static final String MAVEN_REPO = "http://central.maven.org/maven2/";

  @Test
  void test(@TempDir Path root, @TempDir Path javadoc) throws IOException {
    JavadocDownloader downloader = new JavadocDownloader(javadoc, MAVEN_REPO);
    Storage storage = new LocalStorage(root);
    JavadocExtractor extractor = new JavadocExtractor(downloader, storage);

    StepVerifier.create(
            storage.find("com.github.spotbugs", "spotbugs-annotations", "3.1.0-RC3", "index.html"))
        .expectComplete()
        .verify();
    File downloaded =
        extractor
            .extract("com.github.spotbugs", "spotbugs-annotations", "3.1.0-RC3", "index.html")
            .block();
    assertTrue(downloaded.isFile());
    StepVerifier.create(
            storage.find("com.github.spotbugs", "spotbugs-annotations", "3.1.0-RC3", "index.html"))
        .expectNextMatches(file -> file.equals(downloaded))
        .expectComplete()
        .verify();
  }
}

File: src/test/kotlin/jp/skypencil/javadocky/service/JavadocExtractorTest.kt
package jp.skypencil.javadocky.service

import io.kotest.core.spec.style.FunSpec
import io.kotest.engine.spec.tempdir
import io.kotest.matchers.shouldBe
import jp.skypencil.javadocky.repository.LocalStorage
import jp.skypencil.javadocky.repository.Storage
import reactor.test.StepVerifier
import java.io.File

class JavadocExtractorTest : FunSpec({
    val root = tempdir()
    val javadoc = tempdir()

    test("extract a javadoc into storage") {
        val downloader = JavadocDownloader(javadoc.toPath(), MAVEN_REPO)
        val storage: Storage = LocalStorage(root.toPath())
        val extractor = JavadocExtractor(downloader, storage)

        val downloaded = extractor
            .extract("com.github.spotbugs", "spotbugs-annotations", "3.1.0-RC3", "index.html")
            .block()
        downloaded!!.isFile shouldBe true

        StepVerifier.create(
            storage.find("com.github.spotbugs", "spotbugs-annotations", "3.1.0-RC3", "index.html")
        )
            .expectNextMatches { file: File -> file == downloaded }
            .expectComplete()
            .verify()
    }
}) {

    companion object {
        private const val MAVEN_REPO = "https://repo.maven.apache.org/maven2/"
    }
}
